### DISCUSSION TITLE
issue with source maps + typescript + including .ts files from a sibling workspace

### DISCUSSION THREAD

**Discussion by nashspence** [parcel-bundler/parcel#4706] [General] [answered]
Hey! I have been testing out parcel 2 for a few months now, and it has been working pretty well for me. Recently tho, I needed to refactor some of my typescript code into other repository’s. This has changed many of my import paths to now be relative paths that sometimes reference into other workspaces. Since I am in an in-between state, I am not actually including these other workspaces as dependencies, but instead importing the ts source files directly using relative paths. 

So, afaik these files should all be discovered as parcel looks through the imports starting at the entry and properly compiled into the build including source maps.

When using “parcel serve“ however, I see that the code is correctly compiled and bundled, except the source maps for ts files that are multiple directories up the file structure - those starting like “../../../“ - are incorrectly included. 

In Google Chrome, this manifests as the files showing up as sources, however not within the __parcel_source_root folder. Also, when clicking on one of these .ts files I am shown the contents of the entry file index.html. So, seemly the mapping is off.

Right now, I have a work-around where I declare a second entry point as one of the .ts files whose source maps were off. This seems to fix the way the source maps are generated, as they all then show up as expected within the __parcel_source_root folder and the tooling works as expected.

So, it’s not urgent that I get an answer to this, but I am trying to figure out what is causing this problem in the first place. It seems like since I’m including the .ts files directly by relative path they should correctly get source maps no matter where they are. But right now this is not what I’m seeing.

**Comment by gregid** [comment]
I have encountered similar problem with missing source maps in #4699 and shown in the following sample project: https://github.com/gregid/rollup-lib-parcel-app
In my case I haven't succeeded in finding a workaround yet - adding the file as additional entry point doesn't help in my case.

**Reply by nashspence** [reply]
Shoot man. Good to know I’m not alone I suppose. I don’t love the second entry point thing to be honest and was a little surprised that it’s working for me. If you find anything out let me know if you could. If I get a chance to dig into this any more I will maybe try to check out your example and/or file an issue. Right now I’m just not super certain what’s causing the problem and I don’t really have a good repro to submit.

It seems like something must be wrong with the way the source maps are being generated. I say this because the source gets compiled and included in the bundle correctly, but the source maps are seemingly generated incorrectly.

**Reply by gregid** [reply]
Judging by my example it seems like parcel2 doesn't know it should pick up the source map and when to do it.

**Reply by nashspence** [reply]
So like, Babel’s generating them right but they’re getting bundled incorrectly?

**Reply by gregid** [reply]
Yep, that's what I think is happening

**Reply by nashspence** [reply]
That could be it. I honestly forgot to check my bundle and see if I could find the missing maps somewhere in there. If they are there and the paths are just off it might not be a complicated issue

**Comment by mischnic** [comment, marked as answer]
> Also, when clicking on one of these .ts files I am shown the contents of the entry file index.html. So, seemly the mapping is off.

That sounds like Chrome is accessing a file the dev server isn't serving (and instead of 404, it returns the index - as common for SPA apps)

> So, it’s not urgent that I get an answer to this, but I am trying to figure out what is causing this problem in the first place.

This is likely caused by these file being out of the project root (the directory with `.git` or a lockfile):

cc @DeMoorJasper

**Reply by nashspence** [reply]
I did indeed move these files recently into separate repos outside of the original project root. It’s atypical i know, but at the moment it is ideal for me to import these files directly even tho they are not children of the project root. Would you know if this would be something to create an issue about, or if it is by design behavior? Everything other than source maps seems okay with this.

**Reply by mischnic** [reply]
> Would you know if this would be something to create an issue about, or if it is by design behavior?

It's definitely discouraged (also because these files aren't watched, mostly for technical reasons), but it depends on how hard it's to fix

**Reply by nashspence** [reply]
No worries. For what it’s worth, it would be useful for me in my current scenario to have this work as expected. Tho, in the future (once I have things more stable) I will be including these in libraries as dependencies rather than directly and building them separately.

**Reply by DeMoorJasper** [reply]
The serving of `index.html` is definitely due to chrome looking outside of the root, which is prevented for security reasons.

We could inline source files that are outside the project root into the sourcemaps as that would solve this issue.

**Reply by nashspence** [reply]
That seems like a reasonable solution. I wonder if I would be able to configure VSCode to use the inline sources in that case. Alternatively, would it be possible to detect that there were files included outside of the normal project root and have the dev server serve a root that would contain all included files, jic they are not bundled but still need to be accessed like with source maps? No worries if not, I’m not super familiar with this type of thing. I suppose there may be a way for me to change my root accordingly with configuration or whatnot and avoid the problem.
